RFM9x.device.nut

```diff
@@ -0,0 +1,404 @@
+    function setFifoRxBase(start) {
+        _writeReg(RFM9X_REG_FIFO_RX_BASE_ADDR, start & 0xff);
+    }
+    
+    function setCodingRate(cr) {
+        local cur = _readReg(RFM9X_REG_MODEM_CONFIG1);
+        local clear = cur & 0xf1; // exclude bits 3-1
+        _writeReg(RFM9X_REG_MODEM_CONFIG1, clear | (CRTABLE[cr] << 1));
+    }
+    
+    function setRxPayloadCRC(state) {
+        local current = _readReg(RFM9X_REG_MODEM_CONFIG2);
+        _writeReg(RFM9X_REG_MODEM_CONFIG2, (current & 0xfb) | (state ? 0x04 : 0x00));
+    }
+    
+    function setBandwidth(bw) {
+        local current = _readReg(RFM9X_REG_MODEM_CONFIG1);
+        _writeReg(RFM9X_REG_MODEM_CONFIG1, BWTABLE[bw] << 4 | (current & 0x0f));
+    }
+    
+    function setMode(mode) {
+        local current = _readReg(RFM9X_REG_OP_MODE);
+        _writeReg(RFM9X_REG_OP_MODE, mode | (current & 0xf8));
+    }
+        
+    function clearInterrupts() {
+        _writeReg(RFM9X_REG_IRQ_FLAGS, 0xff);
+    }
+    
+    function maskAllInterrupts() {
+        _writeReg(RFM9X_REG_IRQ_FLAGS_MASK, 0xff);
+    }
+    
+    // Enable a specific interrupt. Avaliable interrupts are in
+    // RFM9X_FLAGS
+    function enableInterrupt(mask) {
+        local currentInterrupts = _readReg(RFM9X_REG_IRQ_FLAGS);
+        // Mask the interrupt by writing high, therefore enable by writing low
+        local newInterrupts = currentInterrupts & (~ (1<<mask));
+        _writeReg(RFM9X_REG_IRQ_FLAGS_MASK, newInterrupts);
+    }
+
+    function setMaxPayload(pl) {
+        pl = pl & 0xff;
+        _writeReg(RFM9X_REG_MAX_PAYLOAD_LENGTH, pl);
+    }
+    
+    function setSpreadingFactor(sf) {
+        _writeReg(RFM9X_REG_MODEM_CONFIG2, (sf & 0x0f) << 4);
+    }
+    
+    function setFrequency(freq) {
+        setMode(RFM9X_SLEEP);
+        local reg_freq = (freq/RFM9X_FREQ_STEP).tointeger();
+        local freq_blob = blob(4); // big-endian blob
+        freq_blob[0] = RFM9X_REG_FREQUENCY_HIGH | RFM9X_WRITE_MASK;
+        freq_blob[1] = (reg_freq >> 16) & 0xFF;
+        freq_blob[2] = (reg_freq >> 8) & 0xFF;
+        freq_blob[3] = (reg_freq) & 0xFF;
+        
+        _csLow();
+        _spiModule.writeread(freq_blob);
+        _csHigh();
+    }
+
+    function setReceiveHandler(handler) {
+        _receiveHandler = handler;
+    }
+
+    function isDoneSending() {
+        return _sendFinished;
+    }
+    
+    function _setInLoRaMode() {
+       _writeReg(RFM9X_REG_OP_MODE, RFM9X_CONFIG_BYTE);
+    }
+
+    function _configISR() {
+        if (_irqPin) {
+            _irqPin.configure(DIGITAL_IN_PULLDOWN, _interruptServiceRoutine.bindenv(this));
+        } else {
+            throw "interrupt pin passed to constructor is invalid";
+        }
+    }
+
+    function _interruptServiceRoutine() {
+        // Active high
+        if (_irqPin.read()) {
+            local read = _readReg(RFM9X_REG_IRQ_FLAGS);
+            if (read & (1 << RFM9X_FLAGS.PAYLOAD_CRC_ERROR)) {
+                _error = true;
```

Comment:
The problem with the code is that if an error occurred, user won't be able to send any more messages anymore since the `_sendFinished` is only cleared on success.